auth/gcp/identity.go

```diff
@@ -0,0 +1,202 @@
+	return auth.NewPublicKeySetFromURL(s.cfg.Client, s.cfg.CertURL, time.Hour*2)
+}
+
+// NewIdentityTokenSource will use the GCP metadata services to generate GCP Identity
+// tokens. More information on asserting GCP identities can be found here:
+// https://cloud.google.com/compute/docs/instances/verifying-instance-identity
+func NewIdentityTokenSource(cfg IdentityConfig) (oauth2.TokenSource, error) {
+	if cfg.Client == nil {
+		cfg.Client = &http.Client{
+			Timeout: 5 * time.Second,
+		}
+	}
+	if cfg.MetadataAddress == "" {
+		cfg.MetadataAddress = "http://metadata/computeMetadata/v1/"
+	}
+
+	ts := &idTokenSource{cfg: cfg}
+
+	tok, err := ts.Token()
+	if err != nil {
+		return nil, err
+	}
+	return oauth2.ReuseTokenSource(tok, ts), nil
+}
+
+type idTokenSource struct {
+	cfg IdentityConfig
+}
+
+func (c *idTokenSource) Token() (*oauth2.Token, error) {
+	suffix := fmt.Sprintf("instance/service-accounts/default/identity?audience=%s&format=full",
+		c.cfg.Audience)
+
+	tkn, err := metadataGet(context.Background(), c.cfg, suffix)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to get token")
+	}
+
+	return &oauth2.Token{
+		AccessToken: tkn,
+		TokenType:   "Bearer",
+		Expiry:      timeNow().Add(defaultTokenTTL),
+	}, nil
+}
+
+// IdentityClaimSet holds all the expected values for the various versions of the GCP
+// identity token.
+// More details:
+// https://cloud.google.com/compute/docs/instances/verifying-instance-identity#payload
+// https://developers.google.com/identity/sign-in/web/backend-auth#calling-the-tokeninfo-endpoint
+type IdentityClaimSet struct {
+	jws.ClaimSet
+
+	// Email address of the default service account (only exists on GAE 2nd gen?)
+	Email         string `json:"email"`
+	EmailVerified bool   `json:"email_verified"`
+
+	// Google metadata info (appears to only exist on GCE?)
+	Google map[string]interface{} `json:"google"`
+}
+
+// BaseClaims implements the auth.ClaimSetter interface.
+func (s IdentityClaimSet) BaseClaims() *jws.ClaimSet {
+	return &s.ClaimSet
+}
+
+// IdentityClaimsDecoderFunc is an auth.ClaimsDecoderFunc for GCP identity tokens.
+func IdentityClaimsDecoderFunc(_ context.Context, b []byte) (auth.ClaimSetter, error) {
+	var cs IdentityClaimSet
+	err := json.Unmarshal(b, &cs)
+	return cs, err
+}
+
+// IdentityVerifyFunc auth.VerifyFunc wrapper around the IdentityClaimSet.
+func IdentityVerifyFunc(vf func(ctx context.Context, cs IdentityClaimSet) bool) auth.VerifyFunc {
+	return func(ctx context.Context, c interface{}) bool {
+		ics, ok := c.(IdentityClaimSet)
+		if !ok {
+			return false
+		}
+		return vf(ctx, ics)
+	}
+}
+
+// Issuers contains the known Google account issuers for identity tokens.
+var Issuers = map[string]bool{
+	"accounts.google.com":         true,
+	"https://accounts.google.com": true,
+}
+
+// ValidIdentityClaims ensures the token audience and issuers match expectations.
+func ValidIdentityClaims(cs IdentityClaimSet, audience string) bool {
```

Comment:
This func could probably be a single line `return cs.Aud == audience && Issuers[cs.Iss]`